### PR TITLE
CONTRIBUTING.md: Add contributing.md file

### PR DESCRIPTION
**Pull request by renaynay**
Add current contributing guidelines (these will need updating later in the development process)

### CODE REVIEW

**Review by adlerjohn — DISMISSED** [review]

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+
+Here, `module` is the part of the code that changed (typically a
+top-level Go package), `xxx` is the pull-request number, and `contributor`
+is the author/s of the change.
+
+## Branching Model and Release
+
+The main development branch is main.
+
+Every release is maintained in a release branch named `vX.Y.Z`.
```
Comment (liamsi) [inline CONTRIBUTING.md:67]: This sounds like the release branch is named `vX.Y.Z` while in fact they are (or should be) called `vX.Y` as mentioned below as well. Or am I misunderstanding sth here?

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+Note all pull requests should be squash merged except for merging to a release branch (named `vX.Y`). This keeps the commit history clean and makes it
+easy to reference the pull request where a change was introduced.
+
+### Development Procedure
+
+The latest state of development is on `main`, which must never fail `make test`. _Never_ force push `main`, unless fixing broken git history (which we rarely do anyways).
```
Comment (liamsi) [inline CONTRIBUTING.md:74]: Wait no, we will _never_ force push to master, period.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+To begin contributing, create a development branch on your fork.
+
+Make changes, and before submitting a pull request, update the `CHANGELOG_PENDING.md` to record your change. Also, run either `git rebase` or `git merge` on top of the latest `main`. (Since pull requests are squash-merged, either is fine!)
+
+Sometimes (often!) pull requests get out-of-date with main, as other people merge different pull requests to main. It is our convention that pull request authors are responsible for updating their branches with master. (This also means that you shouldn't update someone else's branch for them; even if it seems like you're doing them a favor, you may be interfering with their git flow in some way!)
```
Comment (liamsi) [inline CONTRIBUTING.md:80]:
```suggestion
Sometimes (often!) pull requests get out-of-date with main, as other people merge different pull requests to main. It is our convention that pull request authors are responsible for updating their branches with `main`. (This also means that you shouldn't update someone else's branch for them; even if it seems like you're doing them a favor, you may be interfering with their git flow in some way!)
```

**Review by liamsi — COMMENTED** [review]
Thank you so much for this!!!

File: CONTRIBUTING.md (hunk 2 above)
Comment (renaynay) [reply CONTRIBUTING.md:74]: I copied this from tendermint sry apparently they force push

File: CONTRIBUTING.md (hunk 1 above)
Comment (renaynay) [reply CONTRIBUTING.md:67]: Screenshot 2022-01-26 at 11 33 43

Comment (liamsi) [reply CONTRIBUTING.md:67]: That is a tag. Not a (necessarily) a branch.

File: CONTRIBUTING.md (hunk 2 above)
Comment (liamsi) [reply CONTRIBUTING.md]: But we don't.

File: CONTRIBUTING.md (hunk 1 above)
Comment (liamsi) [reply CONTRIBUTING.md:67]: I think @Wondertan in the past created a branch as well which was called v0.1.1. But it seems he has deleted that branch as well now? 

Either way: for releases we should create a branch with the name `vX.Y` and on that we tag the releases `vX.Y.0`, `vX.Y.1` and so on.

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+- [module/pkg: Some description about the change #xxx](link to PR) [@contributor](link to contributer github) 
+```
+
+Here, `module` is the part of the code that changed (typically a
+top-level Go package), `xxx` is the pull-request number, and `contributor`
+is the author/s of the change.
+
+## Branching Model and Release
+
+The main development branch is main.
```
Comment (Wondertan) [inline CONTRIBUTING.md:65]:
```suggestion
The main development branch is `main`.
```

File: CONTRIBUTING.md (hunk 1 above)
Comment (Wondertan) [reply CONTRIBUTING.md:67]: Agree with @liamsi

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+Note all pull requests should be squash merged except for merging to a release branch (named `vX.Y`). This keeps the commit history clean and makes it
+easy to reference the pull request where a change was introduced.
+
+### Development Procedure
+
+The latest state of development is on `main`, which must never fail `make test`. _Never_ force push `main`.
+
+To begin contributing, create a development branch on your fork.
+
+Make changes, and before submitting a pull request, update the `CHANGELOG_PENDING.md` to record your change. Also, run either `git rebase` or `git merge` on top of the latest `main`. (Since pull requests are squash-merged, either is fine!)
```
Comment (Wondertan) [inline CONTRIBUTING.md:78]: I think we should make rebase one and only default strategy, as we already follow it. WDYT?

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+Sometimes (often!) pull requests get out-of-date with main, as other people merge different pull requests to main. It is our convention that pull request authors are responsible for updating their branches with `main`. (This also means that you shouldn't update someone else's branch for them; even if it seems like you're doing them a favor, you may be interfering with their git flow in some way!)
+
+#### Merging Pull Requests
+
+It is also our convention that authors merge their own pull requests, when possible. External contributors may not have the necessary permissions to do this, in which case, a member of the core team will merge the pull request once it's been approved.
+
+Before merging a pull request:
+
+- Ensure pull branch is up-to-date with a recent `main` (GitHub won't let you merge without this!)
+- Run `make test` to ensure that all tests pass
```
Comment (Wondertan) [inline CONTRIBUTING.md:89]: Ensure CI green instead?

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+- [Squash](https://stackoverflow.com/questions/5189560/squash-my-last-x-commits-together-using-git) merge pull request
+
+### Git Commit Style
+
+We follow the [Go style guide on commit messages](https://tip.golang.org/doc/contribute.html#commit_messages). Write concise commits that start with the package name and have a description that finishes the sentence "This change modifies Celestia-Node to...". For example,
+
+```sh
+cmd/debug: execute p.Signal only when p is not nil
```
Comment (Wondertan) [inline CONTRIBUTING.md:97]: We should also encourage people to mark the type of the change, e.g. `feat/perf/fix`, etc, like `feat(service/header): new something to do something.` to also follow https://www.conventionalcommits.org/en/v1.0.0/

**Review by liamsi — APPROVED** [review]
👏🏼 🙏🏼 💙 👍🏼

**Review by Wondertan — APPROVED** [review]

**Pull request merged**